File: 2023/src/bin/day2/main.rs
#![warn(clippy::pedantic)]
use std::borrow::Cow;

use advent_2023::{BasicSolution, Solution};
use winnow::{
    ascii::dec_uint,
    combinator::{alt, fold_repeat, opt, preceded, separated_pair},
    Parser,
};

struct Day;

#[derive(Debug, Clone)]
struct Game {
    id: u8,
    revealed: [u8; 3],
}

const RED: usize = 0;
const GREEN: usize = 1;
const BLUE: usize = 2;

impl BasicSolution for Day {
    const INPUT: &'static str = include_str!("data.txt");
    const SAMPLE_INPUT: &'static str = include_str!("sample.txt");

    type Shared = Vec<Game>;
    type Answer = u32;

    const SAMPLE_ANSWER_A: Self::TestAnswer = 8;
    const SAMPLE_ANSWER_B: Self::TestAnswer = 2286;

    fn shared(input: &'static str) -> anyhow::Result<Self::Shared> {
        input
            .lines()
            .map(|line| game.parse(line).map_err(anyhow::Error::msg))
            .collect()
    }

    fn part_a(games: Cow<Self::Shared>) -> anyhow::Result<Self::Answer> {
        Ok(games
            .iter()
            .filter(|game| {
                let [red, green, blue] = game.revealed;
                red <= 12 && green <= 13 && blue <= 14
            })
            .map(|game| u32::from(game.id))
            .sum())
    }

    fn part_b(games: Self::Shared) -> anyhow::Result<Self::Answer> {
        Ok(games
            .into_iter()
            .map(|game| {
                let [red, green, blue] = game.revealed;
                u32::from(red) * u32::from(green) * u32::from(blue)
            })
            .sum())
    }
}

fn game(input: &mut &str) -> winnow::PResult<Game> {
    ("Game ", dec_uint, ": ", revealed)
        .map(|(_, id, _, revealed)| Game { id, revealed })
        .parse_next(input)
}

fn revealed(input: &mut &str) -> winnow::PResult<[u8; 3]> {
    fold_repeat(
        1..,
        preceded(opt(alt((", ", "; "))), color_count),
        || [0; 3],
        |mut acc, (n, color)| {
            acc[color] = acc[color].max(n);
            acc
        },
    )
    .parse_next(input)
}

fn color_count(input: &mut &str) -> winnow::PResult<(u8, usize)> {
    separated_pair(
        dec_uint,
        ' ',
        alt(("red".value(RED), "green".value(GREEN), "blue".value(BLUE))),
    )
    .parse_next(input)
}

fn main() -> anyhow::Result<()> {
    Day::main()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a() -> anyhow::Result<()> {
        Day::test_part_a()
    }

    #[test]
    fn b() -> anyhow::Result<()> {
        Day::test_part_b()
    }
}

File: 2023/python/day2/main.py
import re
from dataclasses import dataclass
from pathlib import Path

SAMPLE_INPUT = """Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green
"""
INPUT = Path("data.txt").read_text()


@dataclass
class Game:
    red: int
    green: int
    blue: int


def part_a(input: str):
    games = [game(line) for line in input.splitlines()]
    return sum(
        i + 1
        for i, game in enumerate(games)
        if game.red <= 12 and game.green <= 13 and game.blue <= 14
    )


def part_b(input: str):
    games = [game(line) for line in input.splitlines()]
    return sum(game.red * game.green * game.blue for game in games)


RED = re.compile(r"(\d+) red")
GREEN = re.compile(r"(\d+) green")
BLUE = re.compile(r"(\d+) blue")


def game(line: str):
    red = max(int(match.group(1)) for match in RED.finditer(line))
    green = max(int(match.group(1)) for match in GREEN.finditer(line))    
    blue = max(int(match.group(1)) for match in BLUE.finditer(line))

    return Game(
        red=red,
        green=green,
        blue=blue,
    )


def test_part_a(capsys):
    assert part_a(SAMPLE_INPUT) == 8
    with capsys.disabled():
        print(f"Part A: {part_a(INPUT)}")


def test_part_b(capsys):
    assert part_b(SAMPLE_INPUT) == 2286
    with capsys.disabled():
        print(f"Part B: {part_b(INPUT)}")
